---
title: "HW1_16s_modified"
output: html_document
---
```{r}
library(dada2); #packageVersion("dada2"); citation("dada2")
library(ShortRead); #packageVersion("ShortRead")
library(ggplot2); #packageVersion("ggplot2")
library(phyloseq); #packageVersion("phyloseq")
```

```{r}
path <- "/Users/Bruce Gan/Desktop/BI586/HW1/16s_sid_rads/Comm_data" # CHANGE ME to the directory containing the fastq files after unzipping.
fns <- list.files(path)
fns
```

```{r}
fastqs <- fns[grepl(".fastq$", fns)]
fastqs <- sort(fastqs)
fnFs <- sort(list.files(path, pattern = "_R1", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "_R2", full.names = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), ".fastq"), `[`, 1)
sample.names2 <- sapply(strsplit(basename(fnRs), ".fastq"), `[`, 1)
```

primer
```{r}
FWD <- "TATGGTAATTGTCTCCTACTTRRSGCAGCAG"
REV <- "AGTCAGTCAGCCGGACTACNVGGGTWTCTAAT"

allOrients <- function(primer) {
  # Create all orientations of the input sequence
  require(Biostrings)
  dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
  orients <- c(Forward = dna, Complement = complement(dna), Reverse = reverse(dna), 
               RevComp = reverseComplement(dna))
  return(sapply(orients, toString))  # Convert back to character vector
}
FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients
REV.orients

fnFs.filtN <- file.path(path, "filtN", basename(fnFs)) # Put N-filterd files in filtN/ subdirectory
fnRs.filtN <- file.path(path, "filtN", basename(fnRs))
filterAndTrim(fnFs, fnFs.filtN, maxN = 0, multithread = TRUE)

primerHits <- function(primer, fn) {
  # Counts number of reads in which the primer is found
  nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
  return(sum(nhits > 0))
}
primerHits
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.filtN[[3]]), 
      FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnFs.filtN[[3]]), 
      REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.filtN[[3]]), 
      REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnFs.filtN[[3]]))
```

```{r}
filt_path <- file.path(path, "trimmed")
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names2, "_R_filt.fastq.gz"))
```

```{r}
#Just to filter out 0 length sequence in order to plot the graph
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen= 1, 
                     maxN=0,
                     maxEE=1,
                     truncQ=2, 
                     trimLeft = 20, 
                     rm.phix=TRUE, 
                     compress=TRUE, multithread=FALSE) 

head(out)
tail(out)
```

```{r}
plotQualityProfile(filtFs[c(1,2,3,4,5,6,7,8,9)])
plotQualityProfile(filtFs[c(10,11,12,13,14,15,16,17,18)])
plotQualityProfile(filtRs[c(1,2,3,4,5,6,7,8,9)])
plotQualityProfile(filtRs[c(10,11,12,13,14,15,16,17,18)])
```

```{r}
#All other settings are default, truncLen is 200 as suggested by the QualityProfile
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen= 200, 
                     maxN=0, 
                     maxEE=1, 
                     truncQ=2, 
                     rm.phix=TRUE, 
                     compress=TRUE, multithread=FALSE) 

head(out)
tail(out)
```

```{r}
errFs <- learnErrors(filtFs, multithread=TRUE)
errRs <- learnErrors(filtRs, multithread=TRUE)
```

```{r}
plotErrors(errFs, nominalQ=TRUE)
```

dereplicates
```{r}
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
names(derepFs) <- sample.names
names(derepRs) <- sample.names2
```

sample inference
```{r}
dadaFs <- dada(derepFs, err=errFs, multithread=TRUE)
dadaRs <- dada(derepRs, err=errRs, multithread=TRUE)
dadaFs[[1]]
dadaRs[[1]]
```

construct sequence table
```{r}
seqtab <- makeSequenceTable(dadaFs)
head(seqtab)
```

merge
```{r}
merger <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
head(merger[[1]])
```

remove chimeras
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```

track reads
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(merger, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

assign taxonomy

```{r}
taxa <- assignTaxonomy(seqtab.nochim, "/Users/Bruce gan/Desktop/BI586/HW1/silva_nr_v132_train_set.fa", multithread=TRUE)
```

```{r}
taxa <- addSpecies(taxa, "/Users/simminishad/Desktop/BI586/16s_sid_rads/silva_species_assignment_v138.fa")
```

```{r}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```
```{r}
write.csv(taxa, file="taxa.csv",row.name=TRUE,quote=FALSE)
unname(head(taxa, 30))
unname(taxa)
```
```{r}
saveRDS(seqtab.nochim, file="final_seqtab_nochim.rds")
saveRDS(taxa, file="final_taxa_blastCorrected.rds")
```
if need to read in previously saved datafiles
```{r}
seqtab.nochim <- readRDS("final_seqtab_nochim.rds")
taxa <- readRDS("final_taxa_blastCorrected.rds")
head(taxa)
```

phyloseq
```{r}
library('phyloseq')
library('ggplot2')
library("Rmisc")
```

```{r}
samdf<-read.csv("mr16s_sampledata.csv",  quote = "", 
                 row.names = NULL, 
                 stringsAsFactors = FALSE)
rownames(samdf) <- samdf$id
head(samdf)
```

```{r}
ids <- paste0("sq", seq(1, length(colnames(seqtab.nochim))))

colnames(seqtab.nochim)<-ids
taxa2 <- cbind(taxa, rownames(taxa)) 
rownames(taxa2)<-ids

ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_names(samdf), 
               tax_table(taxa2))
ps
```
