---
title: "HW1_16s_modified"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r}
library(dada2); #packageVersion("dada2"); citation("dada2")
library(ShortRead); #packageVersion("ShortRead")
library(ggplot2); #packageVersion("ggplot2")
library(phyloseq); #packageVersion("phyloseq")
```

```{r}
path <- "/Users/simminishad/Desktop/BI586/16s_sid_rads/" # CHANGE ME to the directory containing the fastq files after unzipping.
fns <- list.files(path)
fns
```

```{r}
fastqs <- fns[grepl(".fastq$", fns)]
fastqs <- sort(fastqs)
fnFs <- sort(list.files(path, pattern = "_R1", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "_R2", full.names = TRUE))
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
sample.names2 <- sapply(strsplit(basename(fnRs), "_"), `[`, 1)
```

```{r}
filt_path <- file.path(path, "trimmed")
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names2, "_R_filt.fastq.gz"))
```

```{r}
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen= 1, #to filter out 0 length sequences
                     maxN=0, #DADA does not allow Ns
                     maxEE=1, #allow 1 expected errors, where EE = sum(10^(-Q/10)); more conservative, model converges
                     truncQ=2, 
                     trimLeft=20, #N nucleotides to remove from the start of each read: ITS2 primer = F 20bp
                     rm.phix=TRUE, #remove reads matching phiX genome
                     compress=TRUE, multithread=FALSE) # On Windows set multithread=FALSE
head(out)
tail(out)
```

```{r}
plotQualityProfile(filtFs[1])
plotQualityProfile(filtRs[1])
```

```{r}
out <- filterAndTrim(fnFs_rad, filtFr, fnRs_rad, filtRr, truncLen= 200, #to filter out 0 length sequences
                     maxN=0, #DADA does not allow Ns
                     maxEE=1, #allow 1 expected errors, where EE = sum(10^(-Q/10)); more conservative, model converges
                     truncQ=2, 
                     trimLeft=20, #N nucleotides to remove from the start of each read: ITS2 primer = F 20bp
                     rm.phix=TRUE, #remove reads matching phiX genome
                     compress=TRUE, multithread=FALSE) # On Windows set multithread=FALSE
head(out)
tail(out)
```

```{r}
errFs <- learnErrors(filtFs, multithread=TRUE)
errRs <- learnErrors(filtRs, multithread=TRUE)
```
```{r}
plotErrors(errFs, nominalQ=TRUE)
```

dereplicates
```{r}
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
names(derepFs) <- sample.names
names(derepRs) <- sample.names2
```

sample inference
```{r}
dadaFs <- dada(derepFs, err=errFs, multithread=TRUE)
dadaRs <- dada(derepRs, err=errRs, multithread=TRUE)
dadaFs[[1]]
dadaRs[[1]]
```

construct sequence table
```{r}
seqtab <- makeSequenceTable(dadaFs)
head(seqtab)
```

merge
```{r}
merger <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
head(merger[[1]])
```

remove chimeras
```{r}
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```

track reads
```{r}
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(dadaRs, getN), sapply(mergers, getN), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoisedF", "denoisedR", "merged", "nonchim")
rownames(track) <- sample.names
head(track)
```

assign taxonomy

```{r}
taxa <- assignTaxonomy(seqtab.nochim, "/Users/simminishad/Desktop/BI586/16s_sid_rads/silva_nr_v132_train_set.fa", multithread=TRUE)
```

```{r}
taxa <- addSpecies(taxa, "/Users/simminishad/Desktop/BI586/16s_sid_rads/silva_species_assignment_v138.fa")
```

```{r}
taxa.print <- taxa # Removing sequence rownames for display only
rownames(taxa.print) <- NULL
head(taxa.print)
```
```{r}
write.csv(taxa, file="taxa.csv",row.name=TRUE,quote=FALSE)
unname(head(taxa, 30))
unname(taxa)
```
```{r}
saveRDS(seqtab.nochim, file="final_seqtab_nochim.rds")
saveRDS(taxa, file="final_taxa_blastCorrected.rds")
```
if need to read in previously saved datafiles
```{r}
seqtab.nochim <- readRDS("final_seqtab_nochim.rds")
taxa <- readRDS("final_taxa_blastCorrected.rds")
head(taxa)
```


handoff to phyloseq
```{r}
library(phyloseq); packageVersion("phyloseq")
library(Biostrings); packageVersion("Biostrings")
library(ggplot2); packageVersion("ggplot2")
theme_set(theme_bw())
```




